### PR TITLE
Implement SQL UNION

### PR DESCRIPTION
**Pull request by danhhz**
Asking for initial feedback on the design and code structure. The tests need to be expanded and there's quite a few TODOs to work out, but before I went too far down this path, I wanted to make sure it was reasonable.

Does not yet work with ORDER BY or LIMIT on the UNION.

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/4764)

### CODE REVIEW

**Comment by paperstreet** [comment]
Review status: 0 of 3 files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/union.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBoIbvIxNWn0nukdiyw:-KBoIbvIxNWn0nukdiyx:-1868861469) ([raw file](https://github.com/cockroachdb/cockroach/blob/4139fb2cd85dc2e1c2d6e40d44070fe844172be5/sql/union.go#L48)):*
What's supposed to happen in internal errors?

---

*[sql/union.go, line 71 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBoIFZk_QDopSyQWddt:-KBoIFZk_QDopSyQWddu:1971418561) ([raw file](https://github.com/cockroachdb/cockroach/blob/4139fb2cd85dc2e1c2d6e40d44070fe844172be5/sql/union.go#L71)):*
I took this error message from psql. Is that a reasonable thing to do?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764)*

**Comment by RaduBerinde** [comment]
Hi Dan, thanks for working on this! I wrote down some answers/comments, don't consider this a full review.

---

Review status: 0 of 3 files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/union.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBoIbvIxNWn0nukdiyw:-KBoeaIk1LWDjyLrSfNO:-740743061) ([raw file](https://github.com/cockroachdb/cockroach/blob/4139fb2cd85dc2e1c2d6e40d44070fe844172be5/sql/union.go#L48)):*
We can return the error as a query error, or panic. This is fine. BTW you can use a `switch` here.

---

*[sql/union.go, line 71 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBoIFZk_QDopSyQWddt:-KBod5aCLwfPqJ4juoyY:42224916) ([raw file](https://github.com/cockroachdb/cockroach/blob/4139fb2cd85dc2e1c2d6e40d44070fe844172be5/sql/union.go#L71)):*
Definitely. Except that the message won't look ok if this is an INTERSECT or EXCEPT. I would add some tests with this error (you can grep for "pg:" in `testdata` to find other testcases that expect errors).

---

*[sql/union.go, line 148 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBoggyHpT4Dr5itzm7r:-KBoggyHpT4Dr5itzm7s:1434598537) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L148)):*
This is not ok - the stack will be as "deep" as the number of rows. We need to do this stuff in loops.

---

*[sql/union.go, line 164 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBol5dlIqio_PZy6bZw:-KBol5dmZkmJdCjTa311:-1957745055) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L164)):*
this might be easier to follow if we just had two bools `leftDone` and `rightDone`

---

*[sql/union.go, line 183 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBojwstKyt5pbGRxrZX:-KBojwstKyt5pbGRxrZY:1334413981) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L183)):*
Needs a better name. It's a map that stores a count for each encoded tuple; that's what the name should suggest ("emit map" doesn't help in any way").

---

*[sql/union.go, line 185 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBojjxH01mIAUjJ6jeg:-KBojjxH01mIAUjJ6jeh:-1024151106) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L185)):*
These functions need comments and better names. "inc if unseen" is a cumbersome way to say "set to 1". The description and the name should say when it returns true, it shouldn't be tangled up with the logic of "emitting".

---

*[sql/union.go, line 196 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBokaJevJQ1lcuM1dun:-KBokaJevJQ1lcuM1duo:122039661) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L196)):*
why does this return anything if it always returns false

---

*[sql/union.go, line 210 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBofXY-u0z8Px1d4D7N:-KBofXY-u0z8Px1d4D7O:617264727) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L210)):*
This stuff definitely needs more explanation, it is very hard to follow. An example of how the calls "flow" and a mention of what state is maintained in memory in each case would be helpful.

My take is that there is too much abstraction for the amount of logic it "wraps".

I think we could have a single implementation for going through results which is "parameterized" appropriately - one parameter that describes how the map is updated for each row (inc, dec, none, whatever) and one parameter that describes what we return depending on whether the row was found in the map (true if found, true if not found, always true, always false). A union node would use two separate instances of this implementation (left and right), each with appropriate parameters.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-:-KBoms78KP2OOEq_xWlj:1082935325)*

**Comment by petermattis** [comment]
Generally looks good. Fairly clever implementation with the `unionNodeEmit` interface. I agree with Radu that some additional comments describing the structure and flow could be helpful, though I realize that lots of existing code in the `sql` package does not have sufficient commenting. 

---

Review status: 0 of 3 files reviewed at latest revision, 16 unresolved discussions.

---

*[sql/union.go, line 27 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBoyxDF6zQrDCgSMCAM:-KBoyxDF6zQrDCgSMCAN:-1827768261) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L27)):*
`Union.Type` being a string is a historical accident. You should change it to an enumerated type like `sql/parser.BinaryOp` and something like `BinaryOp.String()` to keep formatting working.

---

*[sql/union.go, line 163 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp2GOvrWZkR8c3Mz95:-KBp2GOwBXRLNHQ0a-qs:-770683824) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L163)):*
Not for this PR (or even something to do in the short term), but there is the possibility of emitting the rows immediately if the 2 sides of the union have the same ordering. 

---

*[sql/union.go, line 166 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp-3dhdpE5GTWmHybI:-KBp-3dhdpE5GTWmHybJ:649053325) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L166)):*
I think a `switch` would be easier to read.

---

*[sql/union.go, line 187 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp-pQbBrqOklYjNLdU:-KBp-pQcMfRHJ3QyhLKP:809066962) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L187)):*
You can pass `nil` for `scratch`. That's what you're doing. At some point, it will be good to actually have a scratch buffer that is reused. Might make sense to make `unionNodeEmitMap` a struct to hold that buffer. Or perhaps easier: pass the scratch buffer to `emit{Left,Right}` and down in to these routines.

---

*[sql/union.go, line 196 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBokaJevJQ1lcuM1dun:-KBp0izGP7y4W8TJSCqV:-29241069) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L196)):*
I think it is just convenience. Otherwise the caller always has to return false.

---

*[sql/union.go, line 228 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp08IA6viDFuBajTN_:-KBp08IA6viDFuBajTNa:-955691342) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L228)):*
While I sort of like the grouping you've done with omitting line breaks between some functions, it isn't consistent with the rest of the code base which places a single blank line between functions.

---

*[sql/union.go, line 243 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp4xC-cfxLOFpPrTP9:-KBp4xC0w8yGu2Dqv1ZO:-470328939) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L243)):*
The `string(b)` conversion involves an allocation. Better to store the result to a variable instead of performing it twice. Ditto for the other places this is being done.

---

*[sql/testdata/union, line 34 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp43nLZWMU3OYkE154:-KBp43nMz9a-v1nLKflU:-1044446423) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/testdata/union#L34)):*
Is this correct? I tested with postgres and it returned only `2` for this query which is what I expected.

---

*[sql/testdata/union, line 44 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp3JB914fnPsRdEmEI:-KBp3JB914fnPsRdEmEJ:1374445885) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/testdata/union#L44)):*
Should definitely have some tests that use `SELECT` instead of `VALUES`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-:-KBp5PltxzpV9h6fJbhV:-2002703182)*

**Comment by paperstreet** [comment]
Thanks for the comments! RFAL cc @petermattis @RaduBerinde

Btw, there's a couple places, esp parser/union.go, where we use Union to mean any of the three next to places where we use it to mean UNION specifically. It's confusing and creates some unfortunate naming: Union/UnionType/UnionOp. How do you feel about renaming the generic one to Set in a followup?

---

Review status: 0 of 6 files reviewed at latest revision, 16 unresolved discussions.

---

*[sql/union.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBoIbvIxNWn0nukdiyw:-KBs42U8rFrF9K7U4s6M:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/4139fb2cd85dc2e1c2d6e40d44070fe844172be5/sql/union.go#L48)):*
Done

---

*[sql/union.go, line 71 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBoIFZk_QDopSyQWddt:-KBsk8K-8aFBdjFY74_M:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/4139fb2cd85dc2e1c2d6e40d44070fe844172be5/sql/union.go#L71)):*
Done

---

*[sql/union.go, line 27 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBoyxDF6zQrDCgSMCAM:-KBs40_EuuwxLIXyBhqn:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L27)):*
Done

---

*[sql/union.go, line 148 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBoggyHpT4Dr5itzm7r:-KBs1RqJfGHKlYqhECAW:781682832) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L148)):*
Ha, sorry. Forgot I don't have tail recursion anymore. Fixed.

---

*[sql/union.go, line 163 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp2GOvrWZkR8c3Mz95:-KBs1PRIzzKo9k5wICMc:-1779167650) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L163)):*
Not sure I follow?

---

*[sql/union.go, line 164 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBol5dlIqio_PZy6bZw:-KBs1H0oIlrnkPp6M2wD:-647100554) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L164)):*
Yeah, that's way better. Done

---

*[sql/union.go, line 166 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp-3dhdpE5GTWmHybI:-KBs5gZDtmtEOFiSU1Lw:620070987) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L166)):*
switch doesn't work now that it's split into leftDone and rightDone, but I'll keep this in mind

---

*[sql/union.go, line 183 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBojwstKyt5pbGRxrZX:-KBsZQ3JFZZdIcYGPWs1:1766661308) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L183)):*
Realized that if I special cased the perf optimization directly in readLeft and readRight that this all cleans up considerably. unionNodeEmitMap is gone now

---

*[sql/union.go, line 185 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBojjxH01mIAUjJ6jeg:-KBsZ_MQ7rMJF5bh07_k:-705739638) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L185)):*
Yeah, agreed, though even with better names these were too confusing. I inlined them when I got rid of unionNodeEmitMap

---

*[sql/union.go, line 187 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp-pQbBrqOklYjNLdU:-KBsYePJNPjE9FCF4yPM:1319415913) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L187)):*
Ah, sorry. That's what I assumed that param was for, but the other uses weren't doing it that way. Fixed.

---

*[sql/union.go, line 196 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBokaJevJQ1lcuM1dun:-KBsZwyOKPXCZsAhWUHE:-37173105) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L196)):*
Yeah, that was it

---

*[sql/union.go, line 210 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBofXY-u0z8Px1d4D7N:-KBskGpGe9U8usZhuO4b:1859596569) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L210)):*
I actually originally tried it that way, but found the indirection was more more confusing than helpful. As mentioned in another comment, I realized this all simplifies quite a bit if I special case the one combination (UNION ALL) that doesn't need to compute the hashes. 

---

*[sql/union.go, line 228 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp08IA6viDFuBajTN_:-KBs5pr16TCM49NiCIka:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L228)):*
Done

---

*[sql/union.go, line 243 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp4xC-cfxLOFpPrTP9:-KBs_S5yeEoyf6X7Ncaf:312095602) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L243)):*
This is apparently special cased by the compiler (https://github.com/golang/go/commit/f5f5a8b6209f84961687d993b93ea0d397f5d5bf) and doesn't do an alloc for the string. Added a comment.

---

*[sql/testdata/union, line 34 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp43nLZWMU3OYkE154:-KBs_ehGMdiafKhHnX20:-331886254) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/testdata/union#L34)):*
You are correct, sir. Fixed

---

*[sql/testdata/union, line 44 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp3JB914fnPsRdEmEI:-KBs_dx2_SLAxgh_zI1I:676635175) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/testdata/union#L44)):*
Done

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-:-KBslAC_AoKH_yce9ecI:71114557)*

**Comment by RaduBerinde** [comment]
Review status: 0 of 6 files reviewed at latest revision, 16 unresolved discussions.

---

*[sql/union.go, line 163 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp2GOvrWZkR8c3Mz95:-KBsuSU7yLyhjC5UYtlR:-900856638) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L163)):*
See `planNode.Ordering()` - if the two sides have the outputs sorted in the same way, we can merge the results without building the entire map. E.g. say you have a table `T` with primary index key `K`:
```sql
(SELECT K, V FROM T WHERE V > 5)
UNION
(SELECT K, V FROM T WHERE V < 2)
```
The two selects will return results sorted by `K`, and we can compare them row by row (effectively merging the sorted streams). Like Peter said, not in this change but it would be good to have a TODO or issue filed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764)*

**Comment by tamird** [comment]
Reviewed 1 of 3 files at r1, 4 of 5 files at r3.
Review status: 5 of 6 files reviewed at latest revision, 16 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764)*

**Comment by petermattis** [comment]
LGTM

---

Review status: 5 of 6 files reviewed at latest revision, 11 unresolved discussions.

---

*[sql/union.go, line 163 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp2GOvrWZkR8c3Mz95:-KBt2lXk38KMLaW-oM5l:-1638519687) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L163)):*
Yes, what Radu said. Eventually we could also use `UNION` to hint to the left and right plans that the same sorting would be desirable.

---

*[sql/union.go, line 166 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp-3dhdpE5GTWmHybI:-KBt2_A1bCM7yYhdOThT:-1048233883) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L166)):*
```
switch {
case !n.rightDone:
  return n.readRight()
case !n.leftDone:
  return n.readLeft()
default:
  return false
}
```

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-:-KBt3tvQIFg8h7cCcfxo:676834233)*

**Comment by RaduBerinde** [comment]
> How do you feel about renaming the generic one to Set in a followup?

I wanted to suggest that., maybe `SetOp` or `SetOperation` though since `Set` is very generic

---







*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-:-KBt3zqYkkefOSfTK2qK:-271648454)*

**Comment by paperstreet** [comment]
Okay, I'll bikeshed some names

---

Review status: 5 of 6 files reviewed at latest revision, 11 unresolved discussions.

---

*[sql/union.go, line 163 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp2GOvrWZkR8c3Mz95:-KBt6Exp6_KFxGieC1ys:2136811668) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L163)):*
Ah, gotcha! Added a TODO

---

*[sql/union.go, line 166 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBp-3dhdpE5GTWmHybI:-KBt68QMkWWwIy1zZIjP:1630198576) ([raw file](https://github.com/cockroachdb/cockroach/blob/99923b147c62b03a2c8c6fa5c44c04f988983cc7/sql/union.go#L166)):*
Oh, huh. Okay, done here and elsewhere

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-:-KBt88QI68VxoiSMbnHW:-1662680366)*

**Comment by dt** [comment]
Review status: 5 of 6 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[sql/union.go, line 69 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBtMEjCk6rl2HFsCxhQ:-KBtMEjCk6rl2HFsCxhR:-274331512) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0fc61f2dd853c4111a73bc53fab2201ac92601e/sql/union.go#L69)):*
I think `{r,l}.Hidden` must be equal.

---

*[sql/union.go, line 70 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBtLWEz72oiR1AOTOWJ:-KBtLWEz72oiR1AOTOWK:-662371045) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0fc61f2dd853c4111a73bc53fab2201ac92601e/sql/union.go#L70)):*
nit: s/SQL is/Postgres is/

---

*[sql/union.go, line 126 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBtRrZNh9Sj-04xYCKY:-KBtRrZNh9Sj-04xYCKZ:-1193530750) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0fc61f2dd853c4111a73bc53fab2201ac92601e/sql/union.go#L126)):*
I think you can remove this, and the copy above, and just proxy to `l.Columns`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764)*

**Comment by paperstreet** [comment]
Review status: 5 of 6 files reviewed at latest revision, 12 unresolved discussions.

---

*[sql/union.go, line 69 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBtMEjCk6rl2HFsCxhQ:-KBtXpotPnDWzfd_WVR0:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0fc61f2dd853c4111a73bc53fab2201ac92601e/sql/union.go#L69)):*
Done.

---

*[sql/union.go, line 70 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBtLWEz72oiR1AOTOWJ:-KBtXpS5L3msKvE3iwMh:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0fc61f2dd853c4111a73bc53fab2201ac92601e/sql/union.go#L70)):*
Done.

---

*[sql/union.go, line 126 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBtRrZNh9Sj-04xYCKY:-KBtXrSxzg6wWXFVxWoT:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c0fc61f2dd853c4111a73bc53fab2201ac92601e/sql/union.go#L126)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764)*

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r6.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[sql/union.go, line 25 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBxaC6af_F4AzI5eg6k:-KBxaC6af_F4AzI5eg6l:-1243608516) ([raw file](https://github.com/cockroachdb/cockroach/blob/3594cbc94a750f6d9510d4a7be2f84660f595b51/sql/union.go#L25)):*
how come this function can't return a plain `error`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764)*

**Comment by paperstreet** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[sql/union.go, line 25 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBxaC6af_F4AzI5eg6k:-KBxdLKpGaRBuQrr1ZeQ:-729875512) ([raw file](https://github.com/cockroachdb/cockroach/blob/3594cbc94a750f6d9510d4a7be2f84660f595b51/sql/union.go#L25)):*
So admittedly, I don't have a good sense yet of the difference between error and roachpb.Error, so I was mostly pattern matching. Here's my thoughts:

- makePlan returns a *roachpb.Error, so if this returns an error, it'll have to be converted. Of the existing makePlan entries, the overwhelming majority are not doing this conversion.
- This has to call makePlan on right and left, and do something with the errors, so they'd have to be converted to error and immediately back to roachpb.Error.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764)*

**Comment by tamird** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[sql/union.go, line 25 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764#-KBxaC6af_F4AzI5eg6k:-KBy1MMdG4qaeDAeS-rq:-1833994771) ([raw file](https://github.com/cockroachdb/cockroach/blob/3594cbc94a750f6d9510d4a7be2f84660f595b51/sql/union.go#L25)):*
Ah, the latter is a good reason.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4764)*
